tracepoint_perf/examples/rewrite_perf.rs

```diff
@@ -0,0 +1,281 @@
+        // already populated, and we merge all descriptors with names before merging any
+        // descriptors that don't have names.
+
+        // Prefer data from descriptors that have names.
+        for desc in input.event_desc_list() {
+            if !desc.name().is_empty() {
+                merge_event_desc(&mut output, &output_path, &mut sample_ids_used, desc);
+            }
+        }
+
+        // Fill gaps (if any) using descriptors that don't have names.
+        for desc in input.event_desc_list() {
+            if desc.name().is_empty() {
+                merge_event_desc(&mut output, &output_path, &mut sample_ids_used, desc);
+            }
+        }
+
+        for i in tp::PerfHeaderIndex::FirstFeature.0..tp::PerfHeaderIndex::LastFeature.0 {
+            match tp::PerfHeaderIndex(i) {
+                tp::PerfHeaderIndex::TracingData | tp::PerfHeaderIndex::EventDesc => {
+                    // Let the output file auto-populate these based on add_event_desc and
+                    // add_tracepoint_event_desc.
+                    continue;
+                }
+                index => {
+                    let header_data = input.header(index);
+                    if !header_data.is_empty() {
+                        // Copy the input file's merged header into the output file.
+                        output.set_header(index, header_data);
+                    }
+                }
+            }
+        }
+
+        output.set_tracing_data_required(
+            input.tracing_data_long_size(),
+            input.tracing_data_page_size(),
+        );
+        output.set_tracing_data_header_page(input.tracing_data_header_page());
+        output.set_tracing_data_header_event(input.tracing_data_header_event());
+        output.set_tracing_data_kallsyms(input.tracing_data_kallsyms());
+        output.set_tracing_data_printk(input.tracing_data_printk());
+        output.set_tracing_data_saved_cmd_line(input.tracing_data_saved_cmd_line());
+        output.set_tracing_data_clear_ftraces();
+        for index in 0..input.tracing_data_ftrace_count() {
+            output.set_tracing_data_add_ftrace(input.tracing_data_ftrace(index));
+        }
+
+        if let Err(e) = output.finalize_and_close() {
+            write_error_message(&output_path, e, "output.finalize_and_close failed");
+            _ = fs::remove_file(output_path);
+            continue;
+        }
+
+        println!("{} -> {}", input_path, output_path);
+
+        any_files_converted = true;
+    }
+
+    if any_files_converted {
+        process::ExitCode::SUCCESS
+    } else {
+        process::ExitCode::FAILURE
+    }
+}
+
+fn merge_event_desc(
+    output: &mut tp::PerfDataFileWriter,
+    output_path: &str,
+    sample_ids_used: &mut collections::HashSet<u64>,
+    desc: &td::PerfEventDesc,
+) {
+    let ids = desc.ids();
+    let mut sample_ids_buffer = vec::Vec::with_capacity(ids.len());
```

Comment:
Unnecessary, but this could probably be a thread-local to avoid reallocating for every event.

Reply:
In the C++ version, I had a caller-allocated buffer and passed it in. I decided that wasn't necessary for a sample.